sklearn/utils/_response.py

```diff
@@ -9,6 +9,110 @@
+            (n_samples, n_output)
+        Compressed predictions format as requested by the metrics.
+    """
+    if target_type == "binary" and y_pred.shape[1] < 2:
+        # We don't handle classifiers trained on a single class.
+        raise ValueError(
+            f"Got predict_proba of shape {y_pred.shape}, but need "
+            "classifier with two classes."
+        )
+
+    if target_type == "binary":
+        col_idx = np.flatnonzero(classes == pos_label)[0]
+        return y_pred[:, col_idx]
+    elif target_type == "multilabel-indicator" and isinstance(y_pred, list):
+        # Use a compress format of shape `(n_samples, n_output)`.
+        # Only `MLPClassifier` and `RidgeClassifier` return an array of shape
+        # `(n_samples, n_outputs)`.
+        return np.vstack([p[:, -1] for p in y_pred]).T
+
+    return y_pred
+
+
+def _process_decision_function(*, y_pred, target_type, classes, pos_label):
+    """Get the response values when the response method is `decision_function`.
+
+    This function should process the `y_pred` in the binary and multi-label cases.
+    In the binary case, it should invert the sign of the score if the positive label
+    is not `classes[1]`. In the multi-label case, it should stack the predictions if
+    they are not in the "compressed" format `(n_samples, n_outputs)`.
+
+    Parameters
+    ----------
+    y_pred : ndarray
+        Output of `estimator.predict_proba`. The shape depends on the target type:
+
+        - for binary classification, it is a 1d array of shape `(n_samples,)` where the
+          sign is assuming that `classes[1]` is the positive class;
+        - for multiclass classification, it is a 2d array of shape
+          `(n_samples, n_classes)`;
+        - for multilabel classification, it is a 2d array of shape `(n_samples,
+          n_outputs)`.
+
+    target_type : {"binary", "multiclass", "multilabel-indicator"}
+        Type of the target.
+
+    classes : ndarray of shape (n_classes,) or list of such arrays
+        Class labels as reported by `estimator.classes_`.
+
+    pos_label : int, float, bool or str
+        Only used with binary and multiclass targets.
+
+    Returns
+    -------
+    y_pred : ndarray of shape (n_samples,), (n_samples, n_classes) or \
+            (n_samples, n_output)
+        Compressed predictions format as requested by the metrics.
+    """
+    if target_type == "binary" and pos_label == classes[0]:
+        return -1 * y_pred
+    elif target_type == "multilabel-indicator" and isinstance(y_pred, list):
+        # This is some legacy code where it was assume that the decision function could
+        # be a list of arrays of shape `(n_samples, 2)`.
+        # Currently, this does not exist in scikit-learn. The only estimator supporting
+        # multilabel-indicator and decision_function is `RidgeClassifier` which
+        # returns an array of shape `(n_samples, n_outputs)`.
+        # We could remove this code in the future?
```

Comment:
that means this code is never run? then why do we have it?

Reply:
Because we wrote some tests for it and thus silently supporting it.
We would not have tests, I would have been inclined removing it.
Here, we could potentially break code. It might not be worth it.

Reply:
Then I would simply remove it. It's not a documented behavior anyway.